core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -540,39 +540,56 @@ class ReplicaManager(val config: KafkaConfig,
-        // 2) Record the destination log directory in the memory so that the partition will be created in this log directory
-        //    when broker receives LeaderAndIsrRequest for this partition later.
-        getReplica(topicPartition) match {
-          case Some(_) => // The support for moving replica between log directories on the same broker is not available yet.
-          case None =>
-            logManager.updatePreferredLogDir(topicPartition, destinationDir)
-            throw new ReplicaNotAvailableException(s"Replica $localBrokerId is not available for partition $topicPartition")
-        }
+    replicaStateChangeLock synchronized {
+      partitionDirs.map { case (topicPartition, destinationDir) =>
+        try {
+          if (!logManager.isLogDirOnline(destinationDir))
+            throw new KafkaStorageException(s"Log directory $destinationDir is offline")
+
+          // Stop current replica movement if the destinationDir is different from the existing destination log directory
+          getReplica(topicPartition, Request.FutureLocalReplicaId) match {
+            case Some(futureReplica) =>
+              if (futureReplica.log.get.dir.getParent != destinationDir) {
+                replicaAlterDirManager.removeFetcherForPartitions(Set(topicPartition))
+                logManager.asyncDelete(topicPartition, isFuture = true)
+              }
+            case None =>
+          }
 
-        (topicPartition, Errors.NONE)
-      } catch {
-        case e@(_: LogDirNotFoundException |
-                _: ReplicaNotAvailableException |
-                _: KafkaStorageException) =>
-          (topicPartition, Errors.forException(e))
-        case t: Throwable =>
-          error("Error while changing replica dir for partition %s".format(topicPartition), t)
-          (topicPartition, Errors.forException(t))
+          // If the log for this partition has not been created yet:
+          // 1) Respond with ReplicaNotAvailableException for this partition in the AlterReplicaDirResponse
+          // 2) Record the destination log directory in the memory so that the partition will be created in this log directory
+          //    when broker receives LeaderAndIsrRequest for this partition later.
+          logManager.updatePreferredLogDir(topicPartition, destinationDir)
+          val replica = getReplicaOrException(topicPartition)
+
+          // If the destinationLDir is different from the current log directory of the replica:
+          // 1) If there is no offline log directory, create the future log in the destinationDir.
+          //    Otherwise, return KafkaStorageException. This is needed to prevent future log from being
+          //    created for the same partition in multiple log directories
+          // 2) Otherwise, start ReplicaAlterDirThread to move data of this partition from the current log to the future log
+          if (replica.log.get.dir.getParent != destinationDir) {
```

Comment:
In the current patch, Any will be read and filtered by `ReassignPartitionsCommand` such that broker will not see or handle `any` as the destination log directory. This may change if we want to allow user to cancel ongoing movement. I will reply to the other command after thinking through this.

Reply:
I have replied to the other comment. I think we probably don't need to change AlterReplicaDirRequest or ReassignPartitionsCommand. ReplicaManager will not need to handle the case when destinationDir is Any because AlterReplicaDirRequest is not expected to include "any" as log directory.